### PR TITLE
MCLOUD-6221: Add specific error code in some steps

### PR DESCRIPTION
**Pull request by NadiyaS**
### Description

There are few steps where error codes may be specific with potential solution:

- in step set-crypt-key may be exception thrown FileSystemException. In this case we may add specific exception with code that Crypt key was not set. See src/Step/Deploy/SetCryptKey.php
- src/Step/Deploy/SplitDbConnection.php
- src/Step/Deploy/InstallUpdate/ConfigUpdate/Amqp.php
- src/Step/Deploy/InstallUpdate/ConfigUpdate/CronConsumersRunner.php
- src/Step/Deploy/InstallUpdate/ConfigUpdate/DbConnection.php
- src/Step/Deploy/InstallUpdate/ConfigUpdate/DocumentRoot.php
- src/Step/Deploy/InstallUpdate/ConfigUpdate/Session.php
- src/Step/PostDeploy/Backup.php

### Fixed Issues (if relevant)

1. https://jira.corp.magento.com/browse/MCLOUD-6221

### Manual testing scenarios

1. Create custom scenario where add additional step which change dir permissions to read only:

### Release notes

Added new warning error codes:
2026 - Restoring of some generated data from Build phase to mounted directories was failed.
3004 - Cannot create backup files
More cases were covered to return error code 102 if env.php file is not writable

### Associated documentation updates

Add link to Magento DevDocs PR or Issue, if needed.

### Contribution checklist
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] Pull request introduces user-facing changes and includes meaningful updates for any required release notes and documentation changes
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

File: src/Util/BuildDirCopier.php [hunk 1]
```diff
@@ -78,11 +82,12 @@ public function copy(string $dir, string $strategyName)
                         'Can\'t copy directory %s with strategy: %s',
                         $dir,
                         $strategyName
-                    )
+                    ),
+                    ['errorCode' => Error::WARN_COPY_MOUNTED_DIRS_FAILED]
```
Comment (oshmyheliuk) [inline src/Util/BuildDirCopier.php:86]: This message won't be written in the `cloud.error.log` file as `min_level` for error file handler is set to `warning`.
You can change `$this->logger->notice` to `$this->logger->warning` or decrese level to notice in `\Magento\MagentoCloud\Config\Log::getConfig`

Comment (NadiyaS) [reply src/Util/BuildDirCopier.php:86]: Changed to warning

**Review by oshmyheliuk — DISMISSED** [review]

**Review by bbatsche — DISMISSED** [review]

**Comment by andriyShevtsov** [comment]
QA approved

File: config/schema.error.yaml [hunk 2]
```diff
@@ -565,6 +565,12 @@
     stage: deploy
     step: 'install-update:split-db'
     type: warning
+!php/const Magento\MagentoCloud\App\Error::WARN_COPY_MOUNTED_DIRS_FAILED:
+    title: 'Restoring of some generated data from Build phase to mounted directories was failed.'
```
Comment (meker12) [inline config/schema.error.yaml]:
```suggestion
    title: 'Failed to restore some data generated during the build phase to the mounted directories'
```

File: dist/error-codes.md [hunk 3]
```diff
@@ -145,6 +145,7 @@
 | 2023 | install-update:split-db | Enabling a split database will be skipped. |  |
 | 2024 | install-update:split-db | The SPLIT_DB variable is missing the configuration for split connection types. |  |
 | 2025 | install-update:split-db | Slave connection not set. |  |
+| 2026 | pre-deploy:restore-writable-dirs | Restoring of some generated data from Build phase to mounted directories was failed. | Check the `cloud.log` for more information. |
```
Comment (meker12) [inline dist/error-codes.md]:
```suggestion
| 2026 | pre-deploy:restore-writable-dirs | Failed to restore some data generated data during the build phase to the mounted directories. | Check the `cloud.log` for more information. |
```

File: src/Step/PostDeploy/Backup.php [hunk 4]
```diff
@@ -65,8 +66,15 @@ public function execute()
...
-                $this->file->copy($file, $backup);
-                $this->logger->info(sprintf('Backup %s for %s was created.', $backup, $file));
+                $result = $this->file->copy($file, $backup);
+                if (!$result) {
+                    $this->logger->warning(
+                        sprintf('Backup %s for %s was not created.', $backup, $file),
+                        ['errorCode' => Error::WARN_CREATE_CONFIG_BACKUP_FAILED]
+                    );
+                } else {
+                    $this->logger->info(sprintf('Backup %s for %s was created.', $backup, $file));
```
Comment (meker12) [inline src/Step/PostDeploy/Backup.php]:
```suggestion
                    $this->logger->info(sprintf('Successfully created backup %s for %s .', $backup, $file));
```

File: src/Test/Unit/Util/BuildDirCopierTest.php [hunk 5]
```diff
@@ -191,8 +192,11 @@ public function testCopyWithFilesSystemException(): void
             ->method('getInit')
             ->willReturn($initDir);
         $this->loggerMock->expects($this->once())
-            ->method('notice')
-            ->with('Can\'t copy directory /path/to/root/not-exist-dir. Directory does not exist.');
+            ->method('warning')
+            ->with(
+                'Can\'t copy directory /path/to/root/not-exist-dir. Directory does not exist.',
```
Comment (meker12) [inline src/Test/Unit/Util/BuildDirCopierTest.php]:
```suggestion
                'Cannot copy directory /path/to/root/not-exist-dir. Directory does not exist.',
```

File: src/Test/Unit/Util/BuildDirCopierTest.php [hunk 6]
```diff
@@ -110,7 +111,7 @@ public function copyDataProvider(): array
             ],
             [
                 false,
-                'notice',
+                'warning',
                 'Can\'t copy directory dir with strategy: copy',
```
Comment (meker12) [inline src/Test/Unit/Util/BuildDirCopierTest.php]:
```suggestion
                'Cannot copy directory dir with strategy: copy',
```

File: src/Step/PostDeploy/Backup.php [hunk 7]
```diff
@@ -65,8 +66,15 @@ public function execute()
                 }
 
                 $backup = $file . BackupList::BACKUP_SUFFIX;
-                $this->file->copy($file, $backup);
-                $this->logger->info(sprintf('Backup %s for %s was created.', $backup, $file));
+                $result = $this->file->copy($file, $backup);
+                if (!$result) {
+                    $this->logger->warning(
+                        sprintf('Backup %s for %s was not created.', $backup, $file),
```
Comment (meker12) [inline src/Step/PostDeploy/Backup.php]:
```suggestion
                        sprintf('Failed to create backup %s for %s .', $backup, $file),
```

**Review by oshmyheliuk — DISMISSED** [review]

**Review by bbatsche — APPROVED** [review]

**Review by oshmyheliuk — APPROVED** [review]

**Review by meker12 — APPROVED** [review]

**Pull request merged**